### ISSUE TITLE
[feature request] port build to setuptools

### ISSUE THREAD

**Issue by cdiener** [symengine/symengine.py#156]
Would be awesome if the package could be ported to setuptools as recommended by the Python packaging guide (https://packaging.python.org/guides/tool-recommendations/). This would allow installation with pip and also enable the creation of binary wheels for the major platforms (linux, win and mac).

**Comment by isuruf** [comment]
Installation with pip should work. Please let us know if that doesn't work.

**Comment by cdiener** [comment]
Tried it, but gives me errors in the cake calls. Will post a report tomorrow. Also bdist_wheel does fail as well.

**Comment by cdiener** [comment]
So the problem is that the build arguments (for instance `--symengine-dir`) are not exposed to pip nor to bdist_wheel, so you get:

```shell
pip3 install --user .
Processing /home/cdiener/scribble/symengine.py
Installing collected packages: symengine
  Running setup.py install for symengine ... error
    Complete output from command /usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-d6609kuy-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-_4aamxq2-record/install-record.txt --single-version-externally-managed --compile --user --prefix=:
    /home/cdiener/.local/lib/python3.5/site-packages/setuptools/dist.py:336: UserWarning: Normalizing '0.2.1.dev' to '0.2.1.dev0'
      normalized_version,
    running install
    running build
    running build_ext
    -- The C compiler identification is GNU 6.3.1
    -- The CXX compiler identification is GNU 6.3.1
    -- Check for working C compiler: /usr/bin/cc
    -- Check for working C compiler: /usr/bin/cc -- works
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Check for working CXX compiler: /usr/bin/c++
    -- Check for working CXX compiler: /usr/bin/c++ -- works
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    CMake Error at CMakeLists.txt:5 (find_package):
      Could not find a package configuration file provided by "SymEngine"
      (requested version 0.3.0) with any of the following names:
    
        SymEngineConfig.cmake
        symengine-config.cmake
    
      Add the installation prefix of "SymEngine" to CMAKE_PREFIX_PATH or set
      "SymEngine_DIR" to a directory containing one of the above files.  If
      "SymEngine" provides a separate development package or SDK, be sure it has
      been installed.
    
    
    -- Configuring incomplete, errors occurred!
    See also "/tmp/pip-d6609kuy-build/build/lib.linux-x86_64-3.5/CMakeFiles/CMakeOutput.log".
    error: error calling cmake
    
    ----------------------------------------
Command "/usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-d6609kuy-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-_4aamxq2-record/install-record.txt --single-version-externally-managed --compile --user --prefix=" failed with error code 1 in /tmp/pip-d6609kuy-build/
```

 I tried to run build_ext before that and follow that by pip or bdist_wheel so those would use the cached build but that also fails with:

```shell
cdiener@pest [symengine.py] pip3 install --user .
Processing /home/cdiener/scribble/symengine.py
Installing collected packages: symengine
  Running setup.py install for symengine ... error
    Complete output from command /usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-4oyjj5xq-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-wt8iqt7n-record/install-record.txt --single-version-externally-managed --compile --user --prefix=:
    /home/cdiener/.local/lib/python3.5/site-packages/setuptools/dist.py:336: UserWarning: Normalizing '0.2.1.dev' to '0.2.1.dev0'
      normalized_version,
    running install
    running build
    running build_ext
    CMake Error: The current CMakeCache.txt directory /tmp/pip-4oyjj5xq-build/build/lib.linux-x86_64-3.5/CMakeCache.txt is different than the directory /home/cdiener/scribble/symengine.py/build/lib.linux-x86_64-3.5 where CMakeCache.txt was created. This may result in binaries being created in the wrong place. If you are not sure, reedit the CMakeCache.txt
    CMake Error: The source "/tmp/pip-4oyjj5xq-build/CMakeLists.txt" does not match the source "/home/cdiener/scribble/symengine.py/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.
    error: error calling cmake
    
    ----------------------------------------
Command "/usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-4oyjj5xq-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-wt8iqt7n-record/install-record.txt --single-version-externally-managed --compile --user --prefix=" failed with error code 1 in /tmp/pip-4oyjj5xq-build/
```

whereas bdist_wheels fails with a different error:

```shell
cdiener@pest [symengine.py] python3 setup.py bdist_wheel
/home/cdiener/.local/lib/python3.5/site-packages/setuptools/dist.py:336: UserWarning: Normalizing '0.2.1.dev' to '0.2.1.dev0'
  normalized_version,
running bdist_wheel
running build
running build_ext
SymEngine_DIR : ../symengine/build
-- Python include path: /usr/include/python3.5m
-- Python version: 3.5
-- Python install path: /usr/lib/python3.5/site-packages
-- Found CYTHON: cython
CMAKE_BUILD_TYPE        : Release
CMAKE_CXX_FLAGS_RELEASE : -std=c++11 -Wall -Wextra -fno-common -fPIC -O3 -ffast-math -funroll-loops -Wno-unused-parameter -march=native -fopenmp
CMAKE_CXX_FLAGS_DEBUG   : -std=c++11 -Wall -Wextra -fno-common -fPIC -g -Wno-unused-parameter -ggdb -fopenmp
HAVE_SYMENGINE_MPFR     : False
HAVE_SYMENGINE_MPC      : False
HAVE_SYMENGINE_PIRANHA  : False
HAVE_SYMENGINE_FLINT    : False
HAVE_SYMENGINE_LLVM     : False
Copying source of python wrappers into: /home/cdiener/scribble/symengine.py/build/lib.linux-x86_64-3.5
-- Configuring done
-- Generating done
-- Build files have been written to: /home/cdiener/scribble/symengine.py/build/lib.linux-x86_64-3.5
[100%] Built target symengine_wrapper
installing to build/bdist.linux-x86_64/wheel
running install
running install_egg_info
running egg_info
creating symengine.egg-info
writing symengine.egg-info/PKG-INFO
writing dependency_links to symengine.egg-info/dependency_links.txt
writing top-level names to symengine.egg-info/top_level.txt
writing manifest file 'symengine.egg-info/SOURCES.txt'
reading manifest file 'symengine.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching '*.hpp' under directory 'symengine'
warning: no files found matching '*.cmake' under directory 'symengine'
warning: no files found matching '*.pxi' under directory 'symengine'
warning: no files found matching '*.in' under directory 'cmake'
writing manifest file 'symengine.egg-info/SOURCES.txt'
Copying symengine.egg-info to build/bdist.linux-x86_64/wheel/symengine-0.2.1.dev0-py3.5.egg-info
running install_scripts
SymEngine_DIR : ../symengine/build
-- Python include path: /usr/include/python3.5m
-- Python version: 3.5
-- Python install path: build/bdist.linux-x86_64/wheel/symengine-0.2.1.dev0.data/platlib
-- Found PYTHON: /usr/lib64/libpython3.5m.so  
-- Found CYTHON: cython
CMAKE_BUILD_TYPE        : Release
CMAKE_CXX_FLAGS_RELEASE : -std=c++11 -Wall -Wextra -fno-common -fPIC -O3 -ffast-math -funroll-loops -Wno-unused-parameter -march=native -fopenmp
CMAKE_CXX_FLAGS_DEBUG   : -std=c++11 -Wall -Wextra -fno-common -fPIC -g -Wno-unused-parameter -ggdb -fopenmp
HAVE_SYMENGINE_MPFR     : False
HAVE_SYMENGINE_MPC      : False
HAVE_SYMENGINE_PIRANHA  : False
HAVE_SYMENGINE_FLINT    : False
HAVE_SYMENGINE_LLVM     : False
Copying source of python wrappers into: /home/cdiener/scribble/symengine.py/build/lib.linux-x86_64-3.5
-- Configuring done
-- Generating done
-- Build files have been written to: /home/cdiener/scribble/symengine.py/build/lib.linux-x86_64-3.5
[100%] Built target symengine_wrapper
Install the project...
-- Install configuration: "Release"
CMake Error at symengine/cmake_install.cmake:36 (file):
  file cannot create directory:
  /usr/local/build/bdist.linux-x86_64/wheel/symengine-0.2.1.dev0.data/platlib/symengine.
  Maybe need administrative privileges.
Call Stack (most recent call first):
  cmake_install.cmake:37 (include)


Makefile:117: recipe for target 'install' failed
gmake: *** [install] Error 1
error: error installing
```

which is weird since bdist_wheel shoud install into the wheel (so into ./build or ./dist).

**Comment by isuruf** [comment]
For `bdist_wheel` to work, you need the following patch
```
diff --git a/setup.py b/setup.py
index 118156f..431b8ba 100644
--- a/setup.py
+++ b/setup.py
@@ -168,9 +168,9 @@ class InstallWithCmake(_install):
             cmake_opts.extend(self.define)
 
         cmake_build_type[0] = self.build_type
-        cmake_opts.extend([('PYTHON_INSTALL_PATH', self.install_platlib)])
+        cmake_opts.extend([('PYTHON_INSTALL_PATH', path.join(os.getcwd(), self.install_platlib))])
         cmake_opts.extend([('PYTHON_INSTALL_HEADER_PATH',
-                            self.install_headers)])
+                            path.join(os.getcwd(), self.install_headers))])
 
     def cmake_install(self):
         source_dir = path.dirname(path.realpath(__file__))
```

As for using pip I think you can give build options to pip. Let me get back to you on that

**Comment by cdiener** [comment]
Thanks! Is that patch scheduled for a merge or does it conflict with something else? If I can build a wheel I would already be pretty happy :smile:

**Comment by isuruf** [comment]
Yes, I'll send a PR.

**Comment by cdiener** [comment]
Thanks so much! That was really quick :+1:

**Comment by cdiener** [comment]
On a similar note. do you have plans to put symengine.py on PyPI?

**Comment by isuruf** [comment]
:smile: No-one can possibly be that fast. I fixed it this morning and was going to send a PR later.

You mean wheels or source? Source is already on PyPI

**Comment by cdiener** [comment]
wheels, preferably for all major platforms (linux, mac and windows).

**Comment by isuruf** [comment]
Do you know what the license implications are when we distribute LGPL libraries modified using auditwheel or delocate? Can symengine.py remain MIT licensed?

**Comment by cdiener** [comment]
I would say since you link to it without modifying it should be fine.

**Comment by cdiener** [comment]
[This here](https://stackoverflow.com/questions/104968/can-i-use-lgpl-library-in-bsd-x11-licensed-project/106415#106415) seems to imply that you can if you mention where to get the source or how to compile by yourself. If it's only a few dependencies you could also ask the owner directly.

**Comment by isuruf** [comment]
`pip install symengine --global-option=build_ext --global-option="--symengine-dir=/path/to/symengine"` should send the `symengine-dir` argument to `build_ext`

**Comment by isuruf** [comment]
Wheels are available on PyPI for 0.3.0.rc0 now. Please reopen this issue or a new issue if anything comes up.

**Comment by cdiener** [comment]
Thanks so much!
